Fix MSVC VST compilation

lukas-w opened the pull request:
This PR fixes VST compilation with MSVC (see #3999). Changes made include:

1. A lot of changes to CMake scripts had to be made. `RemoteVstPlugin` is now built using `ExternalProject_Add` when the target platform doesn't match, and is named `RemoteVstPlugin32` and `RemoteVstPlugin64` for 32bit and 64bit respectively. WineGCC isn't used directly any more, but is passed to the CMake external project with `-DCMAKE_CXX_COMPILER=...`-
2. Use [`GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) instead of the hand-written `export.h`. The `EXPORT` define was renamed to `LMMS_EXPORT`.
3. Support for [vcpkg ](https://github.com/Microsoft/vcpkg) and numerous other fixes were added by @justnope in #4352. Thank you and please come back! :heart: 
4. I cherry-picked @Reflexe's SDL2 fixes from #3947 to make LMMS usable. Thanks!
5. `RemoteVstPlugin.cpp` was ported to use C++11 threads instead of POSIX threads. [mingw-std-threads](https://github.com/meganz/mingw-std-threads) is included as a submodule and used for MinGW versions that don't support C++11 threads natively.
6. Vestige reads DLL's file headers to determine whether to use `RemoteVstPlugin32` or `RemoteVstPlugin64`. Thanks @tresf for the [suggestion](https://github.com/LMMS/lmms/issues/4103#issuecomment-357972162).

As a positive side-effect, this enables 64bit VSTs for Linux (fixes #4103). Here's an AppImage: https://transfer.sh/LFwDB/lmms-1.2.0-rc5.137-linux-x86_64.AppImage

tresf commented on .circleci/config.yml:
```diff
@@ -100,7 +100,7 @@ jobs:
             cd build
             make install
-            make appimage
+            make appimage || cat appimage.log
```
👍

PhysSong replied:
Won't this mark the build as successful if AppImage creation is failed? I think it's fine if desired though.

tresf replied:
Perhaps `make appimage || cat appimage.log && false`

PhysSong replied:
> Perhaps `make appimage || cat appimage.log && false`

I think there should be a pair of parenthesis:  `make appimage || (cat appimage.log && false)`
See also: https://unix.stackexchange.com/questions/88850/

lukas-w replied:
I think if appimage creation fails, it will fail one line later in `cp ./lmms-*.AppImage /tmp/artifacts/`. I'll still add the `&& false` to make sure.

PhysSong replied:
> I think if appimage creation fails, it will fail one line later in `cp ./lmms-*.AppImage /tmp/artifacts/`.

True, I just missed that.

tresf commented on cmake/linux/winegcc_wrapper.in:
```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
+# Wrapper script for winegcc to remove .exe file ending automatically
+# appended by winebuild.
+# Usage: winegcc <args ...>
```
Your usage says `winegcc`.  I'm sure this was intentional, but more accurately...

```bash
# Usage: winegcc_wrapper <args ...>

# --- OR ---
# Usage:
#    CONFIGURE_FILE("path/to/winegcc_wrapper.in winegcc_wrapper @ONLY)
#    SET(WINEGCC "path/to/winegcc_wrapper")
#    winegcc <args ...>
```

Probably nitpicking, but didn't know if it was a typo or intentional. :)

lukas-w commented:
I'll be merging this shortly if there are no objections. This changes a lot of things in CMake, so I it's possible it breaks the build for some Linux users. Feel free to complain to me if that's the case.

PhysSong commented on plugins/vst_base/VstPlugin.cpp:
```diff
@@ -91,20 +147,32 @@ VstPlugin::VstPlugin( const QString & _plugin ) :
-		m_badDllFormat = false;
-		tryLoad( "32/RemoteVstPlugin32" );
+	case PE::MachineType::amd64:
+		tryLoad( "RemoteVstPlugin64" );
```
If I understand correctly, this will try to load 64bit DLL in 32bit system using nonexistent `RemoteVstPlugin64`. It will print an error message, but Windows users can't see that.
Do we need to try this on 32bit LMMS?

lukas-w replied:
This is correct, there's no visible error message in the GUI when loading fails. This PR doesn't change that.

PhysSong replied:
I think error message issue can be fixed later, but do we ever need to try to load VST with `RemoteVstPlugin64` on 32bit systems?

lukas-w replied:
Theoretically, we could ship a `RemoteVstPlugin64` with our 32bit Windows installers for users who install 32bit LMMS because they're unaware of the difference. In that case, it's good to try `RemoteVstPlugin64`. In any other case, there's no harm done here. `tryLoad` will discover that `RemoteVstPlugin64` doesn't exist and will abort, at which point we can easily add an error message later.

PhysSong replied:
Got it, I didn't think about that.

PhysSong replied:
I think we may try to detect processor in runtime if possible, though it looks like out of scope.
Qt provides [`QSysInfo::currentCpuArchitecture`](http://doc.qt.io/qt-5/qsysinfo.html#currentCpuArchitecture) as of Qt 5.4, and it's also possible to use OS-dependent detection.

lukas-w replied:
I opened #4447 to track it.

lukas-w commented:
Anything left to change @PhysSong or should I merge?

PhysSong commented on cmake/modules/FindWine.cmake:
```diff
@@ -24,6 +31,10 @@ STRING(REPLACE " " ";" WINEBUILD_FLAGS "${WINEBUILD_OUTPUT}")
 	IF("${FLAG}" MATCHES "libwinecrt0.a.*")
+		STRING(REGEX REPLACE "/wine/libwinecrt0.a.*" "" FLAG "${FLAG}")
+
+		SET(WINE_64_LIBRARY_DIR "${FLAG}/")
```
Unusual, but it is possible to use 32-bit WINE toolchain `wine32-tools` on 64-bit Debian/Ubuntu. In that case, 64-bit VST compilation needs the inverse replacing(32-bit toolchain uses 32-bit libraries, even if `-m64` is specified).

lukas-w replied:
Is that something we have supported before and that this PR breaks?

PhysSong commented on plugins/CMakeLists.txt:
```diff
@@ -75,8 +76,8 @@ IF("${PLUGIN_LIST}" STREQUAL "")
 		stk
-		vestige
 		vst_base
+		vestige
```
I think there's no reason to move this.

lukas-w replied:
There is, `vestige` links to `vstbase`, so `vst_base` needs to be processed first. Otherwise, the target doesn't exist when `vestige` tries to link to it.

PhysSong commented on include/stdshims.h:
```diff
@@ -7,8 +7,10 @@
 #include <utility>
 
-#if (__cplusplus >= 201402L)
+#if (__cplusplus >= 201402L || _MSC_VER)
```
It looks VS 2012 incompatible.
https://github.com/LMMS/lmms/pull/4352#discussion_r191983580

PhysSong commented on plugins/vst_base/RemoteVstPlugin.cpp:
```diff
@@ -767,15 +783,12 @@ bool RemoteVstPlugin::load( const std::string & _plugin_file )
-		{
-			sendMessage( IdVstBadDllFormat );
-		}
+		DWORD error	= GetLastError();
```
Minor, but a tab before `=` doesn't look good here.

PhysSong commented on src/core/RemotePlugin.cpp:
```diff
@@ -62,7 +63,10 @@ void ProcessWatcher::run()
 		fprintf( stderr,
 				"remote plugin died! invalidating now.\n" );
+
+#ifndef SYNC_WITH_SHM_FIFO
```
Could you explain why did you add this check?

lukas-w replied:
If you look at the commit that added this change (https://github.com/LMMS/lmms/pull/4421/commits/9db8cbfb31f9b6d3e754887a6280770e3d3e2bcd#diff-b6e06d3d61c47ede770ac49ad9ea8f9e), you can see that I removed an `invalidate()` call guarded by the same check in `processFinished`. I removed it because it prevented receiving messages after the plugin died (related: https://github.com/LMMS/lmms/issues/4103#issuecomment-357974063, #4371, 4fd8ecd7e4084aedbd39900ffe5cfd107ceba1a0). I figured `@jasp00` had a reason to add the check though and simply missed this `invalidate` call, because I couldn't see a point in doing the `invalidate` twice. So I added the checks here because I removed them together with the `invalidate` call in `processFinished`. I'm not sure about this though, it's mostly guesswork. Maybe @DomClark or @Wallacoloo have any insight on this?

PhysSong replied:
Since `SYNC_WITH_SHM_FIFO` is defined on Windows, this check will prevent LMMS from invalidating plugin. There's no other calls to `invalidate` on Windows, so it may cause crashes/hangs when remote plugin crashes.
Also, dropping the call in `processFinished` may result in delaying invalidation because it checks process state every 200ms. I guess call was added for immediate invalidation.
I guess we can somehow work around the delay and/or the IPC issue.

lukas-w replied:
Should I remove the check? Re-adding the call in `processFinished` will render @Justnope's/@Wallacoloo's 4fd8ecd7e4084aedbd39900ffe5cfd107ceba1a0 useless again.

PhysSong replied:
> Should I remove the check?

I think yes. There's no reason to check that here.
> Re-adding the call in `processFinished` (...) 4fd8ecd useless

It does, but can't we work around this? Though it's not necessary for now and might be hard, I don't think it's impossible. I don't want to block merging this because of the call though.

lukas-w replied:
> It does, but can't we work around this?

We probably can, but I have no idea how to do it.

Anyway, I'll revert both changes because this PR doesn't depend on them any more since the `VstBadDllFormat` message was replaced by the PE file detection.

lukas-w commented:
@PhysSong I fixed the white-space error and removed the `invalidate`-related changes so they can be discussed in #4371.

Thank you for reviewing! I'll merge once you approve.

PhysSong commented on plugins/vst_base/RemoteVstPlugin.cpp:
```diff
@@ -675,11 +691,11 @@ void RemoteVstPlugin::init( const std::string & _plugin_file )
 
 
-static void close_check( int fd )
+static void close_check( FILE* fd )
```
Minor, but it's not a file descriptor anymore. Might `fd` be confusing?

PhysSong commented:
@lukas-w For now, I can't find any more issues. Once you address my comments, I'd say let's merge this.
Anyway, I wonder if other Devs such as @DomClark want to review this one.

lukas-w commented:
@PhysSong Fixed the potential AppImage problem and the confusing `FILE*` variable names. Note that the `FILE*` changes in this PR are in conflict with #4401. Feel free to merge your PR first, I can take care of the merge conflicts then.

lukas-w commented:
Rebased to squash some commits and because I failed to attribute ad4c4f0273d827fac8e5c47d18145da020250cd7 to @Reflexe and merged via 846a2afeaa9bc20f160a7cf086c989273505c890.

PhysSong commented on plugins/vst_base/VstPlugin.cpp:
```diff
@@ -66,27 +69,93 @@
+	FileInfo(QString filePath)
+		: m_file(filePath)
+	{
+		m_file.open(QFile::ReadOnly);
```
This file open can fail if the path passed to `VstPlugin` is relative. So I can't load some VSTs...
There should be a path check logic like one in `VstPlugin::tryLoad`.

lukas-w replied:
Fixed via 7ddca851d1b18bbc64b7d1ee15078b0daa8501c9

PhysSong commented on cmake/modules/FindWine.cmake:
```diff
@@ -33,12 +44,19 @@ FOREACH(FLAG ${WINEBUILD_FLAGS})
 		# WineHQ (/opt/wine-stable, /opt/wine-devel, /opt/wine-staging)
 		STRING(REPLACE "/lib64/wine/" "/lib/wine/" FLAG "${FLAG}")
 
```
This logic doesn't work anymore because we drop `/wine/` before it and causes #4501.
Some possible solutions:

- Use a regular expression to match trailing `lib64`
- Remove `/wine/libwinecrt0.a.*` after replacing
- Always replace `lib64` with `lib` if it doesn't break other platforms

lukas-w replied:
Thanks for tracking this down. Fixed with 515fefa6ce734fcc5cf119b8eff8c3c0db65cdfc.
